### PR TITLE
kubeadm: adding integration tests for init

### PR DESCRIPTION
**Pull request by dmmcquay**
**What this PR does / why we need it**: integration tests for kubeadm init focused on valid and invalid discovery tokens

**Special notes for your reviewer**: /cc @luxas @pires
This was taken from Pires's work in https://github.com/kubernetes/kubernetes/pull/40008 . This is just the testing aspect of it

As it stands, these tests will never complete. The reason being is once it passes with a valid discovery token, it will wait until `[apiclient] Created API client, waiting for the control plane to become ready ` which is an infinite poll (https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/master/apiclient.go#L71). 

This is a WIP
**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dmmcquay** [comment]
@luxas wanted to get some direction from you about how you would like this to move forward. Although this is an integration test, it seems like it will require more set up and more likely falls into e2e type tests. 

What are your thoughts @mikedanese ?

**Comment by pires** [comment]
Can we have separate test files for different commands? Having all in one file will work, the problem I faced was when I started adding more files.

Also, this PR should be WIP.

**Comment by dmmcquay** [comment]
@luxas ping

**Comment by pires** [comment]
@dmmcquay again, can we have separate test files for different commands? Having all in one file will work, the problem I faced was when I started adding more files and that makes sense to split stuff like `kubeadm join --discovery token://` from `kubeadm ex token generate`.

File: cmd/kubeadm/test/token_test.go [hunk 1]
```diff
@@ -65,3 +65,35 @@ func TestCmdTokenGenerateTypoError(t *testing.T) {
...
+		{"--discovery=token://Abcdef:1234567890abcdef", false}, // invalid token non-lowercase
+		{"--discovery=token://abcdef:1234567890abcdef", true},  // valid token
+		{"", true}, // no token provided, so generate
+	}
+
+	for _, rt := range initTest {
+		_, _, actual := RunCmd(kubeadmPath, "init", rt.args, "--skip-preflight-checks")
+		if (actual == nil) != rt.expected {
+			t.Errorf(
+				"failed CmdInitToken running 'kubeadm init %s' with an error: %v\n\texpected: %t\n\t  actual: %t",
```
Comment (luxas) [inline cmd/kubeadm/test/token_test.go]: CmdInit

File: cmd/kubeadm/test/token_test.go [hunk 2]
```diff
@@ -65,3 +65,35 @@ func TestCmdTokenGenerateTypoError(t *testing.T) {
...
+// kubeadmReset executes "kubeadm reset" and restarts kubelet.
+func kubeadmReset() error {
+	_, _, err := RunCmd(kubeadmPath, "reset")
+	return err
+}
+
+func TestCmdInitToken(t *testing.T) {
+	var initTest = []struct {
+		args     string
+		expected bool
```
Comment (luxas) [inline cmd/kubeadm/test/token_test.go]: I can't tell from the name if this is an expected error or an expected success, please make it more clear

**Comment by luxas** [comment]
Are these tests even run? Also, having separate files per command tested makes sense

File: cmd/kubeadm/test/token_test.go (hunk 1 above)
Comment (dmmcquay) [inline cmd/kubeadm/test/token_test.go]: Its the name of the test, which would make the logging more clear.

File: cmd/kubeadm/test/token_test.go (hunk 2 above)
Comment (dmmcquay) [inline cmd/kubeadm/test/token_test.go]: Its how almost all the table driven tests are done currently in kubeadm. I think it is pretty clear. If it doesn't equal the expected value, it should fail.

**Comment by dmmcquay** [comment]
@luxas @pires I will look into making it separate files, but before that, I wanted to have a discussion about integration tests for tests that will poll and wait forever. It doesn't seem possible in the current framework to run something like `kubeadm init` because it will wait for the control plane to become ready. 

Does this mean for these types of tests we will only test failures so as not to hang on the poll?  I think that might be a short term solution but we need to have tests for these.

**Comment by dmmcquay** [comment]
@luxas these tests are run: https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/test-dockerized.sh#L58

**Comment by dmmcquay** [comment]
@luxas have you given any more thought to how you want these integration tests to run?

**Comment by dmmcquay** [comment]
@luxas still waiting to hear back what you want to do about this.  I know it is somewhat attached to #40175 , but that should be wrapping up soon.

**Comment by luxas** [comment]
Let's wait for your other PR first, you have to redo this one regardless...

**Comment by dmmcquay** [comment]
@luxas what I want on this PR is to start the conversation about how we are going to do integration tests. 

If you look at my previous comments, we have to decide what we are going to do with commands that require some preparation before integration tests are run (For example, with `kubeadm init` it expects the control plan). 

What do you think we should do?

**Comment by dmmcquay** [comment]
@luxas I was thinking maybe it would be a good idea to get this pushed through, and then I can move these new tests and the previous integration tests wherever we decide to put them in this https://github.com/kubernetes/kubernetes/pull/40175 . What do you think?

**Comment by pires** [comment]
I think we should wait for a call on how/where to do E2E/integration testing.

**Comment by luxas** [comment]
Let's wait for the conclusion from your doc, then we can return to this

**Comment by dmmcquay** [comment]
@luxas I added the changes that we discussed during the SIG meeting to these changes as well. Should be good to go, PTAL.

**Comment by dmmcquay** [comment]
@k8s-bot cvm gce e2e test this

**Comment by dmmcquay** [comment]
@luxas tests are green.

**Comment by luxas** [comment]
/lgtm
/approve


@dmmcquay @pipejakob Are these cmd tests running in unit/integration presubmit now?

**Comment by pipejakob** [comment]
@luxas Yup. One of the presubmit jobs runs all unit/cmd/integration tests. So long as these run as part of `make test-cmd`, breaking them will block PRs:

https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/test-dockerized.sh#L59

**Pull request merged**